internal/expr/scan.go

```diff
@@ -0,0 +1,196 @@
+			return [][]any{}, err
+		} else if !ok {
+			break
+		}
+
+		rs := []any{}
+		for _, t := range ts {
+			rp := reflect.New(t)
+			// We need to unwrap the struct inside the interface{}.
+			r := rp.Elem()
+			err := re.decodeValue(r)
+			if err != nil {
+				return [][]any{}, err
+			}
+			rs = append(rs, r.Interface())
+		}
+
+		s = append(s, rs)
+	}
+
+	re.Close()
+	return s, nil
+}
+
+// Next calls sql.rows.Next() and checks if it has thrown an error.
+func (re *ResultExpr) Next() (bool, error) {
+	ok := re.rows.Next()
+	if !ok && re.rows.Err() != nil {
+		return false, re.rows.Err()
+	}
+	return ok, nil
+}
+
+// Decode copies the columns in the current row into the fields of the structs in dests.
+// dests can contain any subset of the structs mentioned in the query.
+func (re *ResultExpr) Decode(dests ...any) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot decode expression: %s", err)
+		}
+	}()
+
+	for _, dest := range dests {
+		if dest == nil {
+			return fmt.Errorf("need valid sturct, got nil")
+		}
+
+		v := reflect.ValueOf(dest)
+		if v.Kind() != reflect.Pointer {
+			return fmt.Errorf("need pointer to struct, got non-pointer")
+		}
+		v = reflect.Indirect(v)
+
+		if v.Kind() != reflect.Struct {
+			return fmt.Errorf("need struct, got %s", v.Kind())
+		}
+
+		if err := re.decodeValue(v); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (re *ResultExpr) scan() ([]any, error) {
+	cols, err := re.rows.Columns()
+	if err != nil {
+		return []any{}, fmt.Errorf("cannot scan: %s", err)
+	}
+
+	// rows.Scan takes a list of pointers to where it should scan the values
+	// from the result row.
+	// We call rows.Scan(ptrs) and the results are stored in vals.
+	vals := make([]any, len(cols))
+	ptrs := make([]any, len(cols))
+	for i, _ := range cols {
+		ptrs[i] = &vals[i]
+	}
+	re.rows.Scan(ptrs...)
+
+	// offset is the distace between column X in the row and the position of the
+	// column: "_sqlair_X". It allows non-sqlair columns to be returned in
+	// the results.
+	offset := 0
+
+	sqlairVals := []any{}
+
+	for i, col := range cols {
+		if col == "_sqlair_"+strconv.Itoa(i-offset) {
+			sqlairVals = append(sqlairVals, vals[i])
+		} else {
+			offset++
+		}
+	}
+
+	return sqlairVals, nil
+}
+
+// decodeValue sets the fields in the reflected struct "v" which have tags
+// corrosponding to columns in current row of the query results.
+func (re *ResultExpr) decodeValue(v reflect.Value) error {
+	results, err := re.scan()
+	if err != nil {
+		return err
+	}
+
+	typeFound := false
+	for i, field := range re.outputs {
+		if field.structType == v.Type() {
+			typeFound = true
+			err := setValue(v, field, results[i])
+			if err != nil {
+				return fmt.Errorf("struct %s: %s", v.Type().Name(), err)
+			}
+
+		}
+	}
+	if !typeFound {
+		return fmt.Errorf("no output expression of type %s", v.Type().Name())
+	}
+	return nil
+}
+
+func setValue(dest reflect.Value, fInfo field, val any) error {
```

Comment:
Perhaps `fInfo` could be renamed to something like `dest`? Ideally I'd suggest calling it `field`, but I'm not sure shadowing a type name is a good idea

Reply:
This has all dissapered as well :)